### ISSUE TITLE
Relation to RECOLLECT

### ISSUE THREAD

**Issue by charlesdgburns** [charlesdgburns/NM_TinyRNN#19]
We want to look int of the influence of a shared gate on biologically plausible mechanisms.

As a starting place we want to understand RECOLLECT ([van den Berg et al., 2024](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0316453)).

___ 

**Recollect architecture and learning**

**Backprop through time with shared gate**

**RECOLLECT with shared gate**

**Comment by charlesdgburns** [comment]
**Gradient descent with ADAM**

Notation note: here I use the term 'weights', $w_{i,j}$, to refer to the homologue of synapses between two neurons. The set of all weights I denote by $\theta$. 

An important thing to note is that so far we have done gradient descent with ADAM. The original paper [(kingma & ba, 2015)](https://arxiv.org/pdf/1412.6980) is really clearly written, but there's also a [youtube video](https://www.youtube.com/watch?v=IWvTU6swl_E) that explains it in 20 minutes. The gist is that usually gradient descent involves moving each weight in the direction that minimises some loss function. For example if the data is a sequence of targets $y_t$ and the network outputs esimates $\hat{y}_t$ then this could be the mean squared error between target and estimate at time $t$, which we can sum for all times $t$:

$$E_t = MSE(\hat{y}_t,y_t) $$

Where $\hat{y}_t$ can be thought of as function of the parameters at the given timestep: $\theta_t$.

Then standard backprop updates weights, parameterised by $\theta$, by 
$\theta_t = \theta_{t-1} - \alpha\cdot \nabla_\theta E_t(\theta_{t-1}) $
where $\alpha$ is the learning rate for updating weights. There are two problems in basic gradient descent: 

1. Shrinking updates as the gradient gets smaller (when $\alpha$ is fixed). This is solved by momentum, which acts as an exponential moving average of gradients (errors) for the weights. This essentially gives gradient descent a form of memory. 
2. A global learning rate doesn't allow for targeting specific weights. This is solved by adaptive learning rates (per weight), similarly computed with an exponential moving average. 

That's pretty much it. We incorporate this into a new update after defining he first and second-order momentums:

$$ m_t = \beta_1 m_{t-1} +(1-\beta_1)\nabla_\theta E_t(\theta_{t-1}) $$
$$ v_t = \beta_2 v_{t-1} + (1-\beta_2) \nabla_\theta E_t(\theta_{t-1})$$

there's an engineering trick also that scales up $m_t$ and $v_t$ when $t$ is small:
$$ \hat{m_t} = m_t/(1-\beta_1 ^t)$$
where the power of $t$ means we start with a large $\hat{m_t}$ when $t$ is small and then have a larger $m_t$ as the average momentum accumulates. We define $\hat{v_t}$ analogously.

The final part is the update, which now takes the form:

$$ \theta_t = \theta_{t-1} - \alpha \cdot \hat{m_t}/(\sqrt{\hat{v_t}}+\epsilon)$$

The authors comment on $\hat{m_t}/(\sqrt{\hat{v_t}}+\epsilon)$ acting as a gradient signal-to-noise ratio. When close to the optimum, this nears 0, "a form of automatic annealing".
